picker/docs/rofimoji.1
.\" Automatically generated by Pandoc 2.11.3
.\"
.TH "ROFIMOJI" "1" "February 20, 2021" "Version 5.0.0" "Rofi Third-party Add-on Documentation"
.hy
.SH NAME
.PP
\f[B]rofimoji\f[R] - A character (emoji) picker for rofi
.SH SYNOPSIS
.PP
\f[B]rofimoji\f[R] [\f[B]-h\f[R]] [\f[B]\[en]version\f[R]]
[\f[B]\[en]action\f[R]
{\f[I]type\f[R],\f[I]copy\f[R],\f[I]clipboard\f[R],\f[I]unicode\f[R],\f[I]copy-unicode\f[R]}]
[\f[B]\[en]skin-tone\f[R]
{\f[I]neutral\f[R],\f[I]light\f[R],\f[I]medium-light\f[R],\f[I]moderate\f[R],\f[I]dark
brown\f[R],\f[I]black\f[R],\f[I]ask\f[R]}] [\f[B]\[en]files\f[R]
{\f[I]all\f[R],\f[I]FILE\f[R] [\f[I]FILE\f[R] \&...]]}
[\f[B]\[en]prompt\f[R] \f[I]PROMPT\f[R]] [\f[B]\[en]rofi-args\f[R]
\f[I]ROFI_ARGS\f[R]] [\f[B]\[en]max-recent\f[R] \f[I]MAX_RECENT\f[R]]
[\f[B]\[en]clipboarder\f[R] \f[I]CLIPBOARDER\f[R]] [\f[B]\[en]typer\f[R]
\f[I]TYPER\f[R]] [\f[B]\[en]selector\f[R] \f[I]SELECTOR\f[R]]
.SH DESCRIPTION
.PP
Select, insert, or copy Unicode characters like emoji using rofi.
.SH OPTIONS
.TP
-h, \[en]help
Prints brief usage information.
.TP
\[en]version
show program\[cq]s version number and exit
.TP
\[en]action, -a
Possible values: type, copy, clipboard, unicode, copy-unicode
.RS
.PP
Choose what to do with the selected characters: Directly type them with
the \[lq]Typer\[rq], copy them to the clipboard using the
\[lq]Clipboarder\[rq], or insert them indirectly using the clipboard.
\[lq]unicode\[rq] will type the unicode codepoints of the chosen
characters, \[lq]copy-unicode\[rq] will copy it.
.RE
.TP
\[en]skin-tone=\f[I]skin-tone\f[R], -s \f[I]skin-tone\f[R]
Possible values: neutral, light, medium-light, moderate, dark brown,
black, ask
.RS
.PP
Decide on a skin-tone for all supported emojis.
If not set (or set to \[lq]ask\[rq]), you will be asked for each one
.RE
.TP
\[en]files=\f[I]FILE\f[R] [\f[I]FILE\f[R] \&...], -f \f[I]FILE\f[R] [\f[I]FILE\f[R] \&...]
Read characters from this file instead, one entry per line
.TP
\[en]prompt \f[I]PROMPT\f[R], -r \f[I]PROMPT\f[R]
Set rofimoj\[cq]s prompt
.TP
\[en]rofi-args \f[I]ROFI-ARGS\f[R]
A string of arguments to give to the selector.
.TP
\[en]max-recent \f[I]MAX-RECENT\f[R]
Show at most this number of recently used characters (cannot be larger
than 10)
.TP
\[en]clipboarder \f[I]CLIPBOARDER\f[R]
Possible values: xsel, xclip, wl-copy
.RS
.PP
Choose the application to access the clipboard with manually.
.RE
.TP
\[en]typer \f[I]TYPER\f[R]
Possible values: xdotool, wtype
.RS
.PP
Choose the application to type with manually.
.RE
.TP
\[en]selector \f[I]SELECTOR\f[R]
Possible values: rofi, wofi
.RS
.PP
Choose the selector application manually.
Usually \f[C]rofi\f[R], but for Wayland, you may want \f[C]wofi\f[R].
# KEYBINDINGS
.RE
.PP
(optional) Select multiple emoji with shift+enter
.PP
\f[I]enter\f[R] to insert the emoji directly
.PP
\f[I]alt+c\f[R] to copy it to the clipboard
.PP
\f[I]alt+t\f[R] to type it directly
.PP
\f[I]alt+p\f[R] to insert using the clipboard
.PP
\f[I]alt+1\f[R], \f[I]alt+2\f[R] to insert the most recently used
character (alt+2 for the second most recently one etc.)
.PP
\f[I]alt+u\f[R] to insert the Unicode codepoint
.PP
\f[I]alt+i\f[R] to copy the Unicode codepoint to the clipboard
.PP
Please note that wofi does not support keybindings other than
\f[I]enter\f[R].
.SH FILES
.TP
\f[I]\[ti]/.config/rofimoji.rc\f[R]
Per-user configuration file.
.TP
\f[I]/etc/xdg/xdg-i3/rofimoji.rc\f[R]
Global configuration file.
.TP
\f[I]\[ti]/.local/share/rofimoji/recent\f[R]
Stores the recently used characters
.SH CONFIGURATION
.PP
Args that start with \[lq]\[en]\[rq] (eg.
\[en]version) can also be set in a config file.
.PP
Config file syntax allows: key=value, flag=true, stuff=[a,b,c] (for
details, see syntax at
https://github.com/fdw/rofimoji#example-config-file).
If an arg is specified in more than one place, then commandline values
override values from the config file.
.SH WEBSITE
.PP
https://github.com/fdw/rofimoji
.SH AUTHORS
Fabian Winter.


CHANGELOG.md
# [NEXT]
## Added
- There is a new `print` action to just print out the characters. (#75)

# [5.0.0]
## Added
- `rofimoji` can now be called as a modus from rofi! (#44)
- rofimoji now also works with [wofi](https://hg.sr.ht/~scoopta/wofi/). (#53)
- `rofimoji` now has a manpage. (#57)
- There's a new file for [Nerd Font icons](https://www.nerdfonts.com/).

## Changed
- The emojis from 13.1 have been added!
- Instead of several parameters to choose the input method, they have been consolidated into `--action` (`-a`).
- The annotations on emojis don't contain their name. (#59)
- The results are now sorted by rofi, by default based on Levenshtein distance. (#59)
- All entries are now forced to be left aligned, even for right-to-left languages. (#62)
- The dependency on `pyxdg` was removed.
- Some code cleanup. (#56, #58)

# [4.3.0]
## Added
- Added support for Wayland! (#47)
- Added support for `xclip`. (#21)
- You can type and copy the unicode codepoints. (#48)

## Changed
- The order of the characters in all scripts is now the same as the official one.
- All character data has been updated.

# [4.2.0]
## Added
- The most recently characters are now shown in a separate strip and can be selected with shortcuts (`alt+1` etc). (#39)

## Fixed
- Human emojis can now have their own skin tone even if they're part of a larger sequence. (#41)

# [4.1.1]
## Fixed
- Whitespace characters can be inserted.

# [4.1.0]
## Changed
- The extractors have been rewritten: There are now many, *many* supported symbols (all that Unicode offers), but some may have been renamed.

## Added
- A new file for all kinds of maths symbols.

# [4.0.0]
## Changed
- The new emojis from v13.0 are here!
- `rofimoji` is now modular and has an actual `setup.py`, as the emojis are no longer part of the picker code.
- The arguments to insert using the clipboard have been renamed: `--insert-with-clipboard` and `-p`

## Added
- You can ask `rofimoji` to only ever copy the emojis to the clipboard with `--copy-only` (`-c`).
- There are now data files for all of Unicode's scripts.
- A configuration can also be stored in `$XDG_CONFIG_HOME/rofimoji.rc`.

# [3.0.1]
## Fixed
- A race condition with Firefox is now resolved (#23).

# [3.0.0]
## Added
- You can choose a new input method: `rofimoji` can copy your emojis, paste them into the application and restore the previous contents.
- There are now more keywords included so you can find the appropriate emoji more easily.
- You can select a skin tone by default using cli args.
- You can pass arguments to rofi using `--rofi-args`.
- And you can use alternative emoji lists when you provide the `--emoji-file` parameter.

# [2.1.0]
## Changed
- This release is based on the emoji v12, including all these: https://unicode.org/emoji/charts/emoji-versions.html#2019 .
- Renamed meta files to upper case for better visibility.
- Updated dev dependencies to new versions.

# [2.0.1]
## Fixed
- Fix bug when trying to copy multiple emojis. (#6)

# [2.0.0]
## Changed
- Download emoji list from https://unicode.org/emoji/charts-11.0/full-emoji-list.html instead of emojipedia, as that one didn't work at all anymore.
- Skin color selection is now a second step after certain ("human") emojis. Only the neutral version is included in the main list, which accordingly is a lot smaller now.

## Added
- Downloading, parsing and extracting emoji properties from https://unicode.org/Public/emoji//11.0/emoji-data.txt so that we can find "human" emojis for skin color selection.
- A changelog 😁



picker/rofimoji.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import argparse
import enum
import re
import shlex
import sys
from subprocess import run
from typing import List, Tuple

import configargparse

try:
    from picker.clipboarder import Clipboarder
    from picker.typer import Typer
    from picker.selector import Selector
    from picker.paths import *
except ModuleNotFoundError:
    from clipboarder import Clipboarder
    from typer import Typer
    from selector import Selector
    from paths import *


class Rofimoji:
    skin_tone_selectable_emojis = {'☝', '⛹', '✊', '✋', '✌', '✍', '🎅', '🏂', '🏃', '🏄', '🏇', '🏊',
                                   '🏋', '🏌', '👂', '👃', '👆', '👇', '👈', '👉', '👊', '👋', '👌',
                                   '👍', '👎', '👏', '👐', '👦', '👧', '👨', '👩', '👪', '👫', '👬',
                                   '👭', '👮', '👯', '👰', '👱', '👲', '👳', '👴', '👵', '👶', '👷',
                                   '👸', '👼', '💁', '💂', '💃', '💅', '💆', '💇', '💏', '💑', '💪',
                                   '🕴', '🕵', '🕺', '🖐', '🖕', '🖖', '🙅', '🙆', '🙇', '🙋', '🙌',
                                   '🙍', '🙎', '🙏', '🚣', '🚴', '🚵', '🚶', '🛀', '🛌', '🤌', '🤏',
                                   '🤘', '🤙', '🤚', '🤛', '🤜', '🤝', '🤞', '🤟', '🤦', '🤰', '🤱',
                                   '🤲', '🤳', '🤴', '🤵', '🤶', '🤷', '🤸', '🤹', '🤼', '🤽', '🤾',
                                   '🥷', '🦵', '🦶', '🦸', '🦹', '🦻', '🧍', '🧎', '🧏', '🧑', '🧒',
                                   '🧓', '🧔', '🧕', '🧖', '🧗', '🧘', '🧙', '🧚', '🧛', '🧜', '🧝'}

    fitzpatrick_modifiers = {
        '': 'neutral',
        '🏻': 'light skin',
        '🏼': 'medium-light skin',
        '🏽': 'moderate skin',
        '🏾': 'dark brown skin',
        '🏿': 'black skin'
    }

    class Action(enum.Enum):
        TYPE = 'type'
        COPY = 'copy'
        CLIPBOARD = 'clipboard'
        UNICODE = 'unicode'
        COPY_UNICODE = 'copy-unicode'
        STDOUT = 'print'

    fitzpatrick_modifiers_reversed = {" ".join(name.split()[:-1]): modifier for modifier, name in
                                      fitzpatrick_modifiers.items() if name != "neutral"}

    def __init__(self) -> None:
        self.args = self.parse_arguments()
        self.selector = Selector.best_option(self.args.selector)
        self.typer = Typer.best_option(self.args.typer)
        self.clipboarder = Clipboarder.best_option(self.args.clipboarder)
        self.active_window = self.typer.get_active_window()

    def parse_arguments(self) -> argparse.Namespace:
        parser = configargparse.ArgumentParser(
            description='Select, insert or copy Unicode characters using rofi.',
            default_config_files=config_file_locations
        )
        parser.add_argument('--version', action='version', version='rofimoji 5.1.0-SNAPSHOT')
        parser.add_argument(
            '--action',
            '-a',
            dest='action',
            action='store',
            choices=[action.value for action in self.Action],
            default=self.Action.TYPE.value,
            help='How to insert the chosen characters'
        )
        parser.add_argument(
            '--skin-tone',
            '-s',
            dest='skin_tone',
            action='store',
            choices=['neutral', 'light', 'medium-light', 'moderate', 'dark brown', 'black', 'ask'],
            default='ask',
            help='Decide on a skin-tone for all supported emojis. If not set (or set to "ask"), '
                 'you will be asked for each one '
        )
        parser.add_argument(
            '--files',
            '-f',
            dest='files',
            action='store',
            default=['emojis'],
            nargs='+',
            metavar='FILE',
            help='Read characters from this file instead, one entry per line'
        )
        parser.add_argument(
            '--prompt',
            '-r',
            dest='prompt',
            action='store',
            default='😀 ',
            help='Set rofimoj\'s  prompt'
        )
        parser.add_argument(
            '--rofi-args',
            dest='rofi_args',
            action='store',
            default='',
            help='A string of arguments to give to rofi'
        )
        parser.add_argument(
            '--max-recent',
            dest='max_recent',
            action='store',
            type=int,
            default=10,
            help='Show at most this number of recently used characters (cannot be larger than 10)'
        )
        parser.add_argument(
            '--selector',
            dest='selector',
            action='store',
            type=str,
            choices=['rofi', 'wofi'],
            default=None,
            help='Choose the application to select the characters with'
        )
        parser.add_argument(
            '--clipboarder',
            dest='clipboarder',
            action='store',
            type=str,
            choices=['xsel', 'xclip', 'wl-copy'],
            default=None,
            help='Choose the application to access the clipboard with'
        )
        parser.add_argument(
            '--typer',
            dest='typer',
            action='store',
            type=str,
            choices=['xdotool', 'wtype'],
            default=None,
            help='Choose the application to type with'
        )

        parsed_args = parser.parse_args()
        parsed_args.rofi_args = shlex.split(parsed_args.rofi_args)
        parsed_args.action = next(action for action in self.Action if action.value == parsed_args.action)

        return parsed_args

    def standalone(self) -> None:
        returncode, stdout = self.open_main_rofi_window()

        if returncode == 1:
            sys.exit()
        else:
            if 10 <= returncode <= 19:
                characters = self.load_recent_characters(self.args.max_recent)[returncode - 10].strip()
            else:
                self.choose_action_from_return_code(returncode)
                characters = self.process_chosen_characters(stdout.splitlines())

            self.execute_action(characters)

    def mode_show_characters(self) -> None:
        recent_characters = self.format_recent_characters()

        print("\x00markup-rows\x1ftrue\n")
        if len(recent_characters) > 0:
            print(f"\x00message\x1f{recent_characters}")
        print(self.read_character_files())

    def mode_act_on_selection(self, chosen_character: str, returncode: int) -> None:
        if 10 <= returncode <= 19:
            characters = self.load_recent_characters(self.args.max_recent)[returncode - 10].strip()
            self.execute_action(characters)
        else:
            self.choose_action_from_return_code(returncode)
            self.save_selection_to_cache(re.match(r'^(?:\u200e(?! ))?(?P<char>.[^ ]*) .*', chosen_character).group('char'), '')
            self.mode_select_skin_tone('')

    def mode_select_skin_tone(self, processed_character: str) -> None:
        (characters, processed_characters) = self.load_selection_from_cache()
        processed_characters += processed_character.split(' ')[0]

        for character in characters:
            if character not in self.skin_tone_selectable_emojis:
                processed_characters += character
                characters = characters[1:]
            else:
                self.save_selection_to_cache(characters[1:], processed_characters)
                print('\n'.join(
                    character + modifier + " " + self.fitzpatrick_modifiers[modifier]
                    for modifier in self.fitzpatrick_modifiers
                ))
                return

        cache_file_location.unlink()
        self.save_characters_to_recent_file(processed_characters)
        self.execute_action(processed_characters)

    def choose_action_from_return_code(self, return_code: int):
        if return_code == 20:
            self.args.action = self.Action.COPY
        elif return_code == 21:
            self.args.action = self.Action.TYPE
        elif return_code == 22:
            self.args.action = self.Action.CLIPBOARD
        elif return_code == 23:
            self.args.action = self.Action.UNICODE
        elif return_code == 24:
            self.args.action = self.Action.COPY_UNICODE

    def read_character_files(self) -> str:
        return ''.join(self.load_from_file(file_name) for file_name in self.resolve_all_files())

    def resolve_all_files(self) -> List[str]:
        file_names = self.args.files

        if len(file_names) == 1 and file_names[0] == 'all':
            file_names = [file.stem for file in (Path(__file__).parent / "data").glob("*.csv")]
        return file_names

    def load_from_file(self, file_name: str) -> str:
        provided_file = Path(__file__).parent / "data" / f"{file_name}.csv"
        if Path(file_name).expanduser().is_file():
            actual_file_name = Path(file_name).expanduser()
        elif provided_file.is_file():
            actual_file_name = provided_file
        else:
            raise FileNotFoundError(f"Couldn't find file {file_name!r}")

        return actual_file_name.read_text()

    def load_recent_characters(self, max: int) -> List[str]:
        try:
            return recents_file_location.read_text().strip().split('\n')[:max]
        except FileNotFoundError:
            return []

    def format_recent_characters(self) -> str:
        pairings = [f'\u200e{(index + 1) % 10}: {character}' for index, character in
                    enumerate(self.load_recent_characters(self.args.max_recent))]

        return ' | '.join(pairings)

    def open_main_rofi_window(self) -> Tuple[int, str]:
        return self.selector.show_character_selection(
            self.read_character_files(),
            self.format_recent_characters(),
            self.args.prompt,
            self.args.rofi_args
        )

    def process_chosen_characters(self, chosen_characters: List[str]) -> str:
        processed_characters = ''.join(
            self.add_skin_tone(re.match(r'^(?:\u200e(?! ))?(?P<char>.[^ ]*) .*', character).group('char'))
            for character in chosen_characters
        )
        self.save_characters_to_recent_file(processed_characters)

        return processed_characters

    def add_skin_tone(self, character: str) -> str:
        characters_with_skin_tone = []

        for element in character:
            if element in self.skin_tone_selectable_emojis:
                characters_with_skin_tone.append(self.select_skin_tone(element))
            else:
                characters_with_skin_tone.append(element)

        return ''.join(characters_with_skin_tone)

    def select_skin_tone(self, selected_emoji: chr) -> str:
        skin_tone = self.args.skin_tone

        if skin_tone == 'neutral':
            return selected_emoji
        elif skin_tone != 'ask':
            return selected_emoji + self.fitzpatrick_modifiers_reversed[skin_tone]
        else:
            modified_emojis = '\n'.join(
                selected_emoji + modifier + " " + self.fitzpatrick_modifiers[modifier]
                for modifier in self.fitzpatrick_modifiers
            )

            returncode, skin_tone = self.selector.show_skin_tone_selection(
                modified_emojis,
                selected_emoji + '   ',
                self.args.rofi_args
            )

            if returncode == 1:
                return ''

            return skin_tone.split(' ')[0]

    def get_codepoints(self, char: str) -> str:
        return '-'.join(f'{ord(c):x}' for c in char)

    def save_characters_to_recent_file(self, characters: str) -> None:
        max_recent_from_conf = self.args.max_recent

        old_file_name = recents_file_location
        new_file_name = old_file_name.with_name('recent_temp')

        max_recent = min(max_recent_from_conf, 10)

        new_file_name.parent.mkdir(parents=True, exist_ok=True)
        with new_file_name.open('w+') as new_file:
            new_file.write(characters + '\n')

            try:
                with old_file_name.open('r') as old_file:
                    index = 0
                    for line in old_file:
                        if characters != line.strip():
                            if index == max_recent - 1:
                                break
                            new_file.write(line)
                            index += 1

                old_file_name.unlink()
            except FileNotFoundError:
                pass

        new_file_name.rename(old_file_name)

    def append_to_favorites_file(self, characters: str) -> None:
        favorites_file_location.parent.mkdir(parents=True, exist_ok=True)
        with favorites_file_location.open('a+') as file:
            file.write(characters + '\n')

    def execute_action(self, characters: str) -> None:
        if self.args.action == self.Action.TYPE:
            self.typer.type_characters(characters, self.active_window)
        elif self.args.action == self.Action.COPY:
            self.clipboarder.copy_characters_to_clipboard(characters)
        elif self.args.action == self.Action.CLIPBOARD:
            self.clipboarder.copy_paste_characters(characters, self.active_window, self.typer)
        elif self.args.action == self.Action.UNICODE:
            self.typer.type_characters(self.get_codepoints(characters), self.active_window)
        elif self.args.action == self.Action.COPY_UNICODE:
            self.clipboarder.copy_characters_to_clipboard(self.get_codepoints(characters))
        elif self.args.action == self.Action.STDOUT:
            print(characters)

    def save_selection_to_cache(self, characters: str, processed_characters: str) -> None:
        with cache_file_location.open('w+') as file:
            file.write(f'{self.args.action.value}\n{characters}\n{processed_characters}')

    def load_selection_from_cache(self) -> Tuple[str, str]:
        cache = cache_file_location.read_text().split('\n')
        self.args.action = next(action for action in self.Action if action.value == cache[0].strip())
        return cache[1], cache[2]


def main():
    return_value = os.environ.get('ROFI_RETV')

    if return_value is None:
        Rofimoji().standalone()
    elif return_value == "0":
        Rofimoji().mode_show_characters()
    elif not cache_file_location.is_file():
        chosen = sys.argv[-1]
        del sys.argv[-1]
        Rofimoji().mode_act_on_selection(chosen, int(return_value))
    else:
        chosen = sys.argv[-1]
        del sys.argv[-1]
        Rofimoji().mode_select_skin_tone(chosen)


if __name__ == "__main__":
    main()
